### PR TITLE
Implement SystemChrome.getPreferredOrientations

### LINKED ISSUE #95719: Feature: Reading current `DeviceOrientation` configuration
## Use case

Flutter supports setting the devices allowed device orientation by using [`SystemChrome.setPreferredOrientations()`](https://api.flutter.dev/flutter/services/SystemChrome/setPreferredOrientations.html). I would like to read the current orientation in order to do something like:

1. Read & save current configuration
2. Temporarily set a different preferred device orientation
3. Reset to the configuration from step 1

## Proposal

Add an API like 
```dart
List<DeviceOrientation> currentPreferredOrientations =  await SystemChrome.getPreferredOrientations();
```

### PR DESCRIPTION
**Pull request by foreg**
Implement SystemChrome.getPreferredOrientations as requested in #95719. The main idea is to store requested orientations on the framework side, so it would return exactly the same list as the user has set.

Fixes #95719

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by ueman** [comment]
How does this work before `setPreferredOrientation` is initially called?

**Comment by foreg** [comment]
@ueman, it will return `null`. I thought it would be a better solution than returning an empty list, because in the case mentioned in the issue the library could decide not to invoke `setPreferredOrientation` at all

**Comment by ueman** [comment]
Then a round trip of saving the current orientations to setting some others to restoring the previously saved orientation wouldn't get you where you started. Thus, it's not really solving my problem as outlined in the issue. I prefer the implementation in https://github.com/flutter/engine/pull/39763

**Comment by foreg** [comment]
If you don't mind, could you please clarify why it wouldn't get us where we started?

For example,
Case A:
1) An app sets `portraitUp` - we save it
2) A module gets `portraitUp` and saves it internally
3) The module sets `portraitUp` and `portraitDown` - we save it
4) On exit the module sets `portraitUp`, because it saved that internally

Case B:
1) An app sets an empty list - we save it
2) A module gets the empty list, stores it
3) the same as case A
4) On exit the module sets an empty list

Case C:
1) An app never invokes the set method - we hold null
2) A module gets null, stores it
3) the same as case A
4) On exit the module doesn't invoke the set method

In all of these cases, we will set the same preferred orientations as they were in step 1

File: packages/flutter/lib/src/services/system_chrome.dart [hunk 1]
```diff
@@ -381,6 +381,14 @@ abstract final class SystemChrome {
...
       _stringify(orientations),
     );
+    _preferredOrientations = orientations;
+  }
+
+  /// Returns the list of device orientations as they were requested in [setPreferredOrientations].
+  ///
+  /// Returns null, if the [setPreferredOrientations] was never called.
+  static List<DeviceOrientation>? getPreferredOrientations() {
+    return _preferredOrientations;
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/system_chrome.dart:391]: Don't we have to query the platform here to find out what the current orientations are? Somebody may have changed these on the native side without the framework knowing.

**Comment by ueman** [comment]
> If you don't mind, could you please clarify why it wouldn't get us where we started?

It's mostly the same reason as @goderbauer mentioned. It's very frequent, at least on Android, to set some orientations on the native side. So if you never set it in Flutter, you can read the currently allowed orientations and thus you can't restore it.

I filed the original issue mostly out of the perspective of a library author to do a temporary lock of the orientation, which means I read the current orientations always before setting them. Which is just not possible with this approach. 

I agree though that this is better than before, but ultimately it has to be implemented on the native side. Especially, if you also consider add2app scenarios where the orientation might be locked before a Flutter screen is opened.

The documentation should definitely document the current limitations if/when this gets merged.

**Comment by goderbauer** [comment]
As implemented, this will be confusing because what `getPreferredOrientations` returns may not reflect reality. We'll have to find a different way to implement this.

**Pull request closed without merging**